extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetryableS3OutputStream.java

```diff
@@ -199,15 +211,47 @@ private void pushCurrentChunk() throws IOException
   {
     currentChunk.close();
     final Chunk chunk = currentChunk;
-    try {
-      if (chunk.length() > 0) {
-        resultsSize += chunk.length();
+    if (chunk.length() > 0) {
+      try {
+        SEMAPHORE.acquire(); // Acquire a permit from the semaphore
```

Comment:
Between the static semaphore, the executor threads and other locks, the code has become very difficult to follow and clearly error prone.

If we are trying to avoid running out of disk space, then instead of having semaphores, we just need to check some condition in `write()` that there is disk space available (this can be computed based on the formula that @cryptoe has suggested). If there isn't, we should wait for the condition to be satisfied before proceeding with the write. 

There should be a javadoc in `write()` method which also calls this out clearly.
The class level javadoc might need to be updated too based on the changes here.

Reply:
> If we are trying to avoid running out of disk space, then instead of having semaphores, we just need to check some condition in write() that there is disk space available (this can be computed based on the formula that @cryptoe has suggested). If there isn't, we should wait for the condition to be satisfied before proceeding with the write.

Do you suggest adding a retry wrapper over a new synchronized method assertEnoughDiskSpaceExists()? If yes, for what duration? I feel this would be more complicated to follow and more error prone, and would potentially increase the overall query duration a lot because of a bunch of retries across multiple threads, which defeats the purpose of the optimization. Appreciate your thoughts and clarification on this, thanks!

Reply:
No, you can retain the `wait` and `notify` flow, but the condition that you are waiting on becomes the available disk space, which is what we really care about in the first place (unless there is any other condition that we want to wait on).

It could look something like this:

In `write()`:
```
synchronized(diskSpaceAvailableLock) {
  while (availableDiskSpace < sizeToWrite) {
    diskSpaceAvailableLock.wait();
  }
}
```

After upload and delete:
```
synchronized(diskSpaceAvailable) {
   // update availableDiskSpace
  diskspaceAvailableLock.notifyAll();
}
```

Reply:
Interesting, makes sense. Will try this out, thank you!

Reply:
@kfaraz I had a discussion with @cryptoe yesterday about this approach.
While it would theoretically be possible, this approach will get very complicated and error prone to manage considering how our code is structured, since there are way too many edge scenarios that can go wrong.

`write()` method is called concurrently by multiple threads, and we run into edge scenarios like:
```
say chunk size is 5 MB, and disk space available is 10 MB.

thread 1 has written 3 MB
thread 2 has written 3 MB
thread 3 has written 3 MB

Deadlock!!! --- since each thread has written enough data to reach the disk space threshold, but not enough data to reach the chunk size threshold, and hence none of them trigger an upload(), and hence all of them are stuck in deadlock.
```

Based on yesterday's discussion with Karan, we concluded that we can go with `number of chunks` criteria instead of bytes stored, and hence I've implemented that approach via S3UploadConfig class.